lib/units/storage/amazons3.js

```diff
@@ -0,0 +1,197 @@
+    requtil.validate(req, function() {
+        req.checkBody('url').notEmpty()
+      })
+      .then(function() {
+        return putObject(req.params.plugin, req.body.url)
+      })
+      .then(function(obj) {
+        var plugin = req.params.plugin
+        res.status(201)
+          .json({
+            success: true
+          , resource: {
+              date: new Date()
+            , plugin: plugin
+            , id: obj.id
+            , name: obj.name
+            , href: getHref(plugin, obj.id, obj.name)
+            }
+          })
+      })
+      .catch(requtil.ValidationError, function(err) {
+        res.status(400)
+          .json({
+            success: false
+          , error: 'ValidationError'
+          , validationErrors: err.errors
+          })
+      })
+      .catch(function(err) {
+        log.error('Error storing resource', err.stack)
+        res.status(500)
+          .json({
+            success: false
+          , error: 'ServerError'
+          })
+      })
+  })
+
+  app.post('/s/upload/:plugin', function(req, res, next) {
+    var form = new formidable.IncomingForm()
+    var plugin = req.params.plugin
+    Promise.promisify(form.parse, form)(req)
+      .spread(function(fields, files) {
+        var requests = Object.keys(files).map(function(field) {
+          var file = files[field]
+          log.info('Uploaded "%s" to "%s"', file.name, file.path)
+          return putObject(plugin, file.path, file.name)
+            .then(function (obj) {
+              return {
+                field: field
+              , id: obj.id
+              , name: obj.name
+              , temppath: file.path
+              }
+            })
+        })
+        return Promise.all(requests)
+      })
+      .then(function(storedFiles) {
+        res.status(201).json({
+          success: true,
+          resources: (function() {
+            var mapped = Object.create(null)
+            storedFiles.forEach(function(file) {
+              var plugin = req.params.plugin
+              mapped[file.field] = {
+                date: new Date()
+              , plugin: plugin
+              , id: file.id
+              , name: file.name
+              , href: getHref(plugin, file.id, file.name)
+              }
+            })
+            return mapped;
+          })()
+        })
+        return storedFiles
+      })
+      .then(function (storedFiles){
+        storedFiles.forEach(function (file){
+          log.debug('cleaned up: %s', file.temppath)
+          fs.unlink(file.temppath)
+        })
+      })
+      .catch(function(err) {
+        log.error('Error storing resource', err.stack)
+        res.status(500)
+          .json({
+            success: false,
+            error: 'ServerError'
+          })
+      })
+  });
+
+  app.get('/s/blob/:id/:name', function(req, res) {
+    var params = {
+      Key: req.params.id,
+      Bucket: bucket
+    };
+    s3.getObject(params, function(err, data) {
```

Comment:
Mainly formatting changes, but I would prefer

```js
var params = {
  Key: req.params.id
, Bucket: options.bucket
}

s3.getObject(params, function(err, data) {
  if (err) {
    log.error('Unable to retrieve "%s"', path, err)
    res.sendStatus(404)
    return
  }

  res.set({
    'Content-Type': data.ContentType
  })

  res.send(data.Body)
})
```

instead so that it's more consistent with the rest of the app. I changed `bucket` to `options.bucket` (note: change `option.Bucket` to lowercase) so that it's more obvious where the value is coming from.